hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java

```diff
@@ -3013,38 +3013,42 @@ public TenantUserList listUsersInTenant(String tenantId, String prefix)
   }
 
   @Override
-  public OmVolumeArgs getS3Volume(String accessID) throws IOException {
-
-    final String tenantId;
-    try {
-      tenantId = multiTenantManagr.getTenantForAccessID(accessID);
-      // TODO: Get volume name from DB. Do not assume the same. e.g.
-      //metadataManager.getTenantStateTable().get(tenantId)
-      //    .getBucketNamespaceName();
-      final String volumeName = tenantId;
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("Get S3 volume request for access ID {} belonging to tenant" +
-                " {} is directed to the volume {}.", accessID, tenantId,
-            volumeName);
-      }
-      // This call performs acl checks and checks volume existence.
-      return getVolumeInfo(volumeName);
-
-    } catch (OMException ex) {
-      if (ex.getResult().equals(INVALID_ACCESSID)) {
-        // If the user is not associated with a tenant, they will use the
-        // default s3 volume.
-        String defaultS3volume =
-            HddsClientUtils.getDefaultS3VolumeName(configuration);
-
+  public OmVolumeArgs getS3Volume() throws IOException {
+    // Unless the OM request contains S3 authentication info with an access
+    // ID that corresponds to a tenant volume, the request will be directed
+    // to the default S3 volume.
+    String s3Volume = HddsClientUtils.getDefaultS3VolumeName(configuration);
+    S3Authentication s3Auth = getS3Auth();
+
+    if (s3Auth != null) {
+      String accessID = s3Auth.getAccessId();
+      Optional<String> optionalTenantId =
+          multiTenantManager.getTenantForAccessID(accessID);
+
+      if (optionalTenantId.isPresent()) {
+        String tenantId = optionalTenantId.get();
+        s3Volume = metadataManager.getTenantStateTable().get(tenantId)
```

Comment:
Add a null check before this line just in case `metadataManager.getTenantStateTable().get(tenantId)` returns null (in the case of inconsistent OM DB)?

Reply:
Yeah I can add that. What we really need here is a volume lock across the getTenantForAccessID and getTenantStateTable().get() calls. There's little to no locking on any multi-tenant read requests right now, which is why we have HDDS-6063 filed.